packages/datadog-instrumentations/src/graphql.js

```diff
@@ -235,14 +235,20 @@ function callInAsyncScope (fn, thisArg, args, abortController, cb) {
   try {
     const result = fn.apply(thisArg, args)
     if (result && typeof result.then === 'function') {
-      // bind callback to this scope
-      result.then(
-        res => cb(null, res),
-        err => cb(err)
-      )
-    } else {
-      cb(null, result)
+      return new Promise((resolve, reject) => {
+        result.then(
+          res => {
+            cb(null, res)
+            resolve(res)
+          },
+          err => {
+            cb(err)
+            reject(err)
+          }
+        )
+      })
```

Comment:
Rather than a whole new promise object, can't you just return or throw in the `then` handler?

```suggestion
      return result.then(
        res => {
          cb(null, res)
          return res
        },
        err => {
          cb(err)
          throw err
        }
      )
```

That way you're not creating a whole extra promise.

Reply:
Fixed!

Reply:
@bengl looking at this again: in another library, they used the original promise for state tracking.

This is somewhat conflicting with the mongoose implementation, since it means it is impossible to separate the state tracking and prevent calling the promise only once. I guess that is fine overall, it is just something to keep in mind and I wonder if another library could have a similar behavior.

I also wonder if we should discuss with the mongoose maintainers about being spec compliant.

Reply:
Hmm, in that case, should we return the original promise? Like:

```js
result.then(/* .... */)
return result
```

Reply:
This is the current day behavior and is what results in the mongoose error.